M2/Macaulay2/packages/PlaneCurveLinearSeries.m2

```diff
@@ -0,0 +1,1189 @@
+   of |D0'-Dinf'|+B.
+   
+   The routine @TO projectiveImage@ provides the image of the map to projective space given
+   by |D0-Dinf|.  There are special routines for the most important case,
+   @TO canonicalSeries@ and @TO canonicalImage@.
+   
+   The functions @TO addition@ and @TO negative@ implement the group law in the case
+   of a curve of genus 1.
+///
+
+doc ///
+Key
+ addition
+ (addition, List, List, List, Ring)
+ (addition, Ideal, Ideal, Ideal)
+ [addition, Conductor]
+Headline
+ addition of smooth points on a curve of genus 1
+Usage
+ L =  addition(o,p,q,C)
+ I = addition(oC, pC, qC)
+Inputs
+ o: List
+ oC: Ideal
+ p: List
+ pC: Ideal
+ q: List
+ qC: Ideal
+ C: Ring
+Outputs
+ L: List
+ I: Ideal
+Description
+  Text
+   If E is an irreducible plane curve of degree 3 then the set of smooth
+   points of C is a principal homogeneous space under the group Pic_0 E
+   of invertible sheaves of degree 0. Thus if we choose a smooth
+   point o the map p -> O_E(p-o) identifies the set of smooth points with
+   such invertible sheaves. 
+   
+   This script computes the sum of smooth points p,q with respect to the
+   group law in which o is the zero point, where the group operation
+   makes p+q = r if, as divisors, r is linearly equivalent to p+q-o.
+   The functions @TO addition@ and @TO negative@, based on  @TO linearSeries@, 
+   allow us to implement the group law.
+
+   The points o,p,q may be represented either by their homogeneous coordinates
+   or by ideal in the ring E. The functions @TO fromCoordinates@
+   and @TO toCoordinates@ pass between these two representations.
+   
+   Here is an example with a smooth plane cubic:
+  Example
+   kk = QQ
+   S = kk[x,y,z]
+   p = {0,1,0}; pS = fromCoordinates(p,S)
+   q = {1,0,0}; qS = fromCoordinates(q,S)
+   o = {1,1,1}; oS = fromCoordinates(o,S)
+
+   I = ideal random(3, intersect(oS, pS, qS))
+   E = S/I   
+
+   r = addition(o,p,q, E)
+   addition(o, negative(o, p, E), r, E)
+  Text
+   It is known that when one takes multiples of a point that is not torsion,
+   the "height" - roughly the number of digits in the coordinates - is doubled
+   with each iteration, that is, the number of digits doubles:
+  Example
+   q' := o;
+   netList ({q}| for i from 0 to 3 list(
+   q' = addition(o,p,q',E)))
+   
+  Text
+   On the other hand, over a finite field, a curve has only finitely many
+   points, so any subgroup of the Jacobian is finite:
+   
+  Example
+   kk = ZZ/7
+   S = kk[x,y,z]
+   p = {0,1,0}; pS = fromCoordinates(p,S)
+   o = {1,1,1}; oS = fromCoordinates(o,S)
+   setRandomSeed 0
+   I = ideal random(3, intersect(pS,oS))
+   E = S/I   
+   geometricGenus E
+   q = o
+   netList ({o} | apply(5, i-> q = addition(o,p,q,E)))
+  Text
+   A cubic with a node or cusp is also arithmetic genus 1; in the case
+   of a node, the smooth points are in correspondence with P^1 minus {0, infinity}
+   and the Jacobian is the multiplicative troup of the field.
```

Comment:
@DanGrayson I don't understand why the spellchecker didn't run on this PR, but the misspelling of "group" is causing tests to fail on other PRs.

Reply:
The spellchecker didn't run because this PR originally targeted `master` and currently those tests are only run against PR's to `development`.  This shouldn't be a problem once #3080 is merged into `master`.